src/CPUActivityAgent.cpp

```diff
@@ -0,0 +1,532 @@
+        , M_NUM_CORE(m_platform_topo.num_domain(GEOPM_DOMAIN_CORE))
+        , m_do_write_batch(false)
+        , m_update_qm_max_rate(true)
+    {
+        geopm_time(&m_last_wait);
+    }
+
+    // Push signals and controls for future batch read/write
+    void CPUActivityAgent::init(int level, const std::vector<int> &fan_in, bool is_level_root)
+    {
+        m_core_frequency_requests = 0;
+        m_uncore_frequency_requests = 0;
+        m_resolved_f_uncore_efficient = 0;
+        m_resolved_f_uncore_max = 0;
+        m_resolved_f_core_efficient = 0;
+        m_resolved_f_core_max = 0;
+
+        // These are not currently guaranteed to be the system uncore min and max,
+        // just what the user/admin has previously set.
+        m_freq_uncore_min = m_platform_io.read_signal("CPU_UNCORE_FREQUENCY_MIN_CONTROL", GEOPM_DOMAIN_BOARD, 0);
+        m_freq_uncore_max = m_platform_io.read_signal("CPU_UNCORE_FREQUENCY_MAX_CONTROL", GEOPM_DOMAIN_BOARD, 0);
+
+        init_platform_io();
+    }
+
+    void CPUActivityAgent::init_platform_io(void)
+    {
+        for (int domain_idx = 0; domain_idx < M_NUM_CORE; ++domain_idx) {
+            m_core_scal.push_back({m_platform_io.push_signal("MSR::CPU_SCALABILITY_RATIO",
+                                                             GEOPM_DOMAIN_CORE,
+                                                             domain_idx), NAN});
+            m_core_freq_control.push_back({m_platform_io.push_control("CPU_FREQUENCY_MAX_CONTROL",
+                                                                      GEOPM_DOMAIN_CORE,
+                                                                      domain_idx), -1});
+        }
+
+        for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+            m_qm_rate.push_back({m_platform_io.push_signal("MSR::QM_CTR_SCALED_RATE",
+                                                           GEOPM_DOMAIN_PACKAGE,
+                                                           domain_idx), NAN});
+
+            m_uncore_freq_status.push_back({m_platform_io.push_signal("CPU_UNCORE_FREQUENCY_STATUS",
+                                                                      GEOPM_DOMAIN_PACKAGE,
+                                                                      domain_idx), NAN});
+        }
+
+        for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+            m_uncore_freq_min_control.push_back({m_platform_io.push_control("CPU_UNCORE_FREQUENCY_MIN_CONTROL",
+                                              GEOPM_DOMAIN_PACKAGE,
+                                              domain_idx), -1});
+            m_uncore_freq_max_control.push_back({m_platform_io.push_control("CPU_UNCORE_FREQUENCY_MAX_CONTROL",
+                                              GEOPM_DOMAIN_PACKAGE,
+                                              domain_idx), -1});
+        }
+
+        //Configuration of QM_CTR must match QM_CTR config used for training data
```

Comment:
```suggestion
        // Configuration of QM_CTR must match QM_CTR config used for training data
```
Also this comment should explain what is being setup here.  For instance, what is EVENT_ID 2 and why do we care?